clients/src/main/java/org/apache/kafka/common/utils/Utils.java

```diff
@@ -1225,13 +1226,11 @@ public static long tryWriteTo(TransferableChannel destChannel,
      * @param length The number of bytes to write
      * @throws IOException For any errors writing to the output
      */
-    public static void writeTo(DataOutput out, ByteBuffer buffer, int length) throws IOException {
+    public static void writeTo(DataOutputStream out, ByteBuffer buffer, int length) throws IOException {
         if (buffer.hasArray()) {
             out.write(buffer.array(), buffer.position() + buffer.arrayOffset(), length);
         } else {
-            int pos = buffer.position();
-            for (int i = pos; i < length + pos; i++)
-                out.writeByte(buffer.get(i));
+            Channels.newChannel(out).write(buffer);
```

Comment:
- Is this preserving the same behaviour i.e. copy in the range `[pos, pos + length]`? 
- This is mutating the position of the `buffer`, as opposed to the current implementation.

Reply:
1) Yes, per the docs of WritableByteChannel

```
Writes a sequence of bytes to this channel from the given buffer.
An attempt is made to write up to r bytes to the channel, where r is the number of bytes remaining in the buffer, that is, src.remaining(), at the moment this method is invoked.

Suppose that a byte sequence of length n is written, where 0 <= n <= r. This byte sequence will be transferred from the buffer starting at index p, where p is the buffer's position at the moment this method is invoked; the index of the last byte written will be p + n - 1. Upon return the buffer's position will be equal to p + n; its limit will not have changed.
```

2) Good point, although I don't think this has an effect on any of the 4 current users (DefaultRecord.writeTo and LegacyRecord.writeTo for writing key and value), I've added a defensive call to asReadOnlyBuffer.

Reply:
Thanks for the follow-up. 

1. This new implementation ignores the `length` argument provided to the method if the buffer is not backed by an array. What if `length` does not equal the number of remaining bytes on the buffer?

2. Is there an actual optimization offered by calling `write`? The implementation of direct buffers use a similar index-based iteration.

Reply:
Got it. Thanks. 

Would you have performance gains at a high-level, without sharing details on the application?

Publicly available data showing the expectable gains would back up the PR further. Has this been discussed in the dev mailing list already?

Reply:
Sure. In a particular workload, this code path was about 30% of CPU usage in flamegraphs. It is now 2-3% after a local patch.

This hasn't been discussed in dev- its just an attempt to upstream a small performance improvement.

Reply:
Thanks for the PR. Can you please provide a bit more detail regarding the workload? I assume we're talking about the producer here. If so, can you please share whether compression was used, the compression algorithm (if applicable) and the average message size.

Reply:
This is producer code, with no compression (the data is pre-encrypted so it would be useless anyway) and the message size is 1-2 kB

Reply:
ping @ijuma / @Hangleton , are there any blockers to getting this merged?

Reply:
Apologies for the delay - would you have any JMH benchmark for this change? E.g. something like in #13312.

Reply:
Sorry for the delay. I was trying to understand how to show the improvement, but we seem to always pass a heap byte buffer. How can I reproduce this improvement?